### ISSUE TITLE
Mapping destination object problems causes hibernate problems

### ISSUE THREAD

**Issue by dattebayorob** [modelmapper/modelmapper#477]
"Identifier of an instance of was altered"

When mapping a Dto source to a object destination (Hibernate proxy entity) with attributes changing the FK key, hibernates exception happens. If I change to dto source to class destination and set the other attributes manually works normal

Example:
-- Mapper
public void toEntity( Entity entityPersisted, Dto dto) {
		modelMapper.map(dto, entityPersisted, SOME_TYPE_MAP);
}
-- ServiceImpl
repository.findById(dto.getId()),map(entityPersisted -> {
					mapper.toEntity(entityPersisted, dto);
					
					repository.save(entityPersisted);
 }

Caused by: org.hibernate.HibernateException: identifier of an instance of com.lala.project.model.entity.SomeEntityAssociative was altered from 2 to 1

**Comment by dattebayorob** [comment]
there was no option to force the map to generate a new Entity with the dto param. My solution: leave modelmapper

**Comment by zhangalex** [comment]
From DTO/FORM to JPA entity, just use BeanUtils.copyProperties

From Entity to DTO/FORM/VO, just use modelmapper

**Comment by dattebayorob** [comment]
I did it ... But BeanUtils can't map automatically productId to new Product(id), how do you get it to work?

**Comment by zhangalex** [comment]
> I did it ... But BeanUtils can't map automatically productId to new Product(id), how did you get it to work?

create a method like
`
public Product getProduct() {
  if (getProductId() == null) {
     return null;
  }

return new Product(getProductId());
}
`

**Comment by ft0907** [comment]
I also encountered a similar error, how do you solve it?

**Comment by chhsiao90** [comment]
Hi @ft0907 and @dattebayorob ,

Can you provide more detail for your case so I can figure out how to use ModelMapper to fulfill your your case or we can improve ModelMapper, thanks.

**Comment by dattebayorob** [comment]
For example,
On a Entity with properties like this (I will represent it on a json):
```
{
  person: { 
     Id: 1,
     user: { id: 1, name: "some user" }
     profile: { id: 599, level: "MASTER" }
  }
```     
In a PUT cenário all I want when changing the profile property, is to change the FK to a other profile entity.

```
PUT
{ person: {
    id: 1,
    user: { Id: 1 }
    profile: { Id: 33 }
}
```
 
But getting the persisted entity to update with the DTO, modelMapper will map the property instead of instanciate a new entity. Hibernate will not allow it.

**Comment by chhsiao90** [comment]
It's expect behavior to overwrite existing properties with ModelMapper.

It looks like what you want is to merge 2 beans into 1 bean with specific logic (instantiate new instance  if the id is changed), and I guessed ModelMapper is not playing well with this kind of scenario.

Currently, I don't have any good idea to fix this kind of issue. Maybe to not use ModelMapper to merge is the best choice as ModelMapper is used for mapping.

And thanks for reporting the issue. Please let me know if you have any thought! Thanks!

**Comment by dattebayorob** [comment]
Yeah, and sadly there isn't a simple way for conditional mapping. The api for TypeMap leaks of documentation, and after understanding What should be done the code turns almost unreadable.

**Comment by davidrochin** [comment]
No workaround then?

**Comment by vinaykumarnune** [comment]
u you can use model mapper but instead of using same object create a new entity objet with Updated data.it will by default update the entity based on "id".
The main problem was that you are try to change the instance multiple times using model mapper.

**Comment by hanserya** [comment]
@davidrochin Oddly enough, reverting to version 1.1.2 of the model mapper library seems to be a workaround. I have no clue what changed between then and now. 

I was in the middle of an upgrade for a project that I inherited and ran across this gem. Taking the older version made the error go away. I'll try and dig more into it and provide feedback if I get a chance. Unfortunately, I'm already 40 hours into an upgrade that was only supposed to take a quarter of that time so I'll be playing catch up for the next week as a result of this.

@chhsiao90 this doesn't seem to be a problem with model mapper in so far as I can discern. The failure is actually quite logical. My predicament is that the prior dev team (the totality of which was given a chance to pursue "new opportunities a few months ago) has seemingly designed a massive entity/dto-dto/entity mapping "framework" around what appears to be a bug from the prior version of model mapper. If memory serves, then the tool already exposes an API that allows developers to deliberately address these edge cases. I think people got used to the "magic" and are going to start coming by more often now that the magic is gone in the newer version. 

Serves them right for being lazy ;)

**Comment by hanserya** [comment]
Actually; I feel like I've run up against this before. Didn't model mapper use to rely on accessors in order to perform the mapping? I was pretty sure that the newer version uses reflection to manipulate fields on the destination directly. If that is the case, then you have your answer as to why this just started happening. Almost no one places a set method on the identity field for their entity.

**Comment by kirinalexdev** [comment]
For me this problem disappeared when using @Transactional(readOnly = true)
